Bring in improved HAL APIs to master

bulislaw opened the pull request:
### Description

This one PR to rule them all (tm) combines:
https://github.com/ARMmbed/mbed-os/pull/6996
https://github.com/ARMmbed/mbed-os/pull/6995
https://github.com/ARMmbed/mbed-os/pull/6994

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->

@ARMmbed/mbed-os-maintainers 

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] New target
    [x] Feature
    [ ] Breaking change

c1728p9 committed:
Add documentation and test the HAL RTC API

Add requirements, tests, an example implementation and additional
function documentation to the HAL RTC API.

0xc0170 committed:
Sleep: add time requirements for sleep

Sleep - within 10us
Deepsleep - within 10ms

Note about mbed boards with interface, moved to lpc176x, as they are target related,
should be documented in the target documentation.

The tests will come as separate PR, to conform to this updates to sleep API.

c1728p9 committed:
Add documentation for the HAL Ticker API

Add documentation and test header for the HAL Ticker API.

mprse committed:
Add tests for ticker HAL API.

c1728p9 committed:
Turn off RTC until updated

Turn off RTC for all devices. When support for a device is added this
should be re-enabled.

c1728p9 committed:
Keep rtc code if low power timer is used

Keep the RTC code if either DEVICE_RTC or DEVICE_LOWPOWERTIMER is
defined on the devices which use the RTC for both the rtc api and the
low power timer api. This allows DEVICE_LOWPOWERTIMER to be enabled while
DEVICE_RTC is turned off.

c1728p9 committed:
Keep RTC API prototypes even when not enabled

Keep the prototypes in rtc_api.h even when DEVICE_RTC is not defined.
This allows devices that aren't fully compliant with the RTC API to
still use the header and prototypes.

mprse committed:
Add minor modifications of lp/us tickers tests desctiption in ticker header files.

Add one potential bug to the test description.
Fix test function names to be consistant with desctiption and test header.
Fix typos.

mprse committed:
Modify rtc_api.h header file - update tests specification.

mprse committed:
Update of RTC HAL API tests.

Provide minor fixes for existing test cases.
Add additional test cases.

mprse committed:
Add Sleep test specification to sleep_api.h.

mprse committed:
Add test and test header to Sleep HAL API.

bulislaw committed:
Add clarification to rtc_free doxygen

mprse committed:
NRF5 - make us ticker to be driven by high speed 1MHz timer

According to new ticker standards the following requirements for us ticker are not met on RRF5 boards:
- has a frequency between 250KHz and 8MHz (currently is driven by 32kHz clock)
- ticker increments by 1 each tick (currently is scaled to 1 MHz by incrementing counter by ~31)

Since BLE softdevice uses TIMER0 the proposition is to use high speed TIMER1 for us ticker configured as follows:
- TIMER counter width: 16 bits (max)
- TIMER frequency: 1MHz
This solution also uses Timer's capture/compare register 0 to specify interrupt time and Timer's capture/compare register 1 to read current timer value.

mprse committed:
Disable boards which does not fulfil new ticker standards.

mprse committed:
Disable boards which does not fulfil new sleep standards.

These boards will be re-enabled when sleep driver for them is ready.

Note:
This operation is done by removing "SLEEP" feature from target's "device_has" list (in targets.json config file).
For NRF52_DK removing of "SLEEP" feature causes some timing issues which have influence on tests. In order to successfully disable this board we need to disable also related features like "USTICKER", "LOWPOERTIMER" and slightly modify ticker tests, so they will not be executed if usticker support is not available (by default all targets support us ticker).

mprse committed:
Add DEVICE_RTC symbol check in RTC source files for NCS36510 target.

mprse committed:
Skip higher level ticker tests for targets with stale ticker target specific drivers.

Since target specific ticker drivers are not ready also features which uses ticker in upper layers may not work correctly and tests for these features.
We need to disable also failing higher level ticker related tests (by adding check if DEVICE_USTICKER symbol is available and raise #error [NOT_SUPPORTED] if not).

Commit:
Add lp/us ticker free() function implementation for NRF5

mprse committed:
Enable us/lp ticker support for NRF51_DK board

mprse committed:
Modify requirement for the ticker_init function.

Current version:
The function ticker_init resets the internal count and disables the ticker interrupt.

Proposed version:
The function ticker_init allows the ticker to keep counting and disables the ticker interrupt.

This is a result of the following discussion:
https://github.com/ARMmbed/mbed-os/pull/5233#pullrequestreview-86677815

mprse committed:
Modify ticker_init_test() to be consistent with the new requirement.

mprse committed:
Provide fix to the implementation of ticker_overflow_test test case.

The intention was to use ticker_overflow_delta equal to 0 for low power ticker tests and ticker_overflow_delta equal to 50 for high frequency ticker tests. Current implementation is invalid since for devices which provide LOW_POWER_TIMER feature delta is equal to 0 and for devices without this feature delta is equal 50.

mprse committed:
Update lp ticker to be consistent with the new standards.

Provide the following modifications for lp ticker driver:
- According to NRF51_DK reference manual rtc interrupt cannot be controlled by rtc event. In the previous implementation interrupts were enabled permanently and specific interrupt was enabled/disabled by enabling/disabling the specific event. If event is enabled, then event signal is provided to Programmable Peripheral Interconnect (PPI). If interrupt is enabled, then interrupt signal is provided to Nested Vector Interrupt Controller (NVIC). Disable all events permanently. Enable lp ticker overflow interrupt permanently(needed for RTC), disable lp ticker capture/compare interrupt on init (lp_ticker_init) , enable lp ticker interrupt when lp ticker interrupt is set (lp_ticker_set_interrupt), disable lp ticker interrupt on disable request(lp_ticker_disable_interrupt).
- Provide lp ticker data for higher level (freq: 32kHz / len: 24 bits),
- Add the following features to init routine: disable lp ticker interrupt.
- Make ticker driver to operate on ticks instead of us.
- Simplify lp ticker read and set interrupt routines (upper layers handle conversion to us and interrupt scheduling).

Commit:
Add free function to lp/us ticker HAL API

mprse committed:
Enable ticker support for K64F board.

mprse committed:
Adapt K64F us ticker driver to the new standards.

- provide ticker configuration: 1MHz/32 bit counter.
- us_ticker_init() routine disables interrupts.
- adapt comments.

mprse committed:
Adapt K64F lp ticker driver to the new standards.

Low power ticker time counter is created based on RTC which is driven by 32KHz clock. Additional low power timer is used to generate interrupts.
We need to adapt driver to operate on ticks instead of us.

Perform the following updates:
- provide lp ticker configuration: 32KHz/32 bit counter.
- lp_ticker_init() routine disables lp ticker interrupts .
- adapt the driver functions to operate on ticks instead us.
- adapt comments.
- add us_ticker_free() routine.

mprse committed:
lp_us_tickers test - provide counter overflow protection

Since according to the ticker requirements min acceptable counter size is 12 bits (low power timer) for which max count is 4095, then all test cases must be executed in this time window.
HAL ticker layer handles overflow and it is not handled in the target ticker drivers.

mprse committed:
lp_us_tickers test - add tolerance to interrupt time.

On some platforms (e.g. K64F) different counters are used for time measurement and interrupt generation.
Because of that we should relax interrupt test case and give additional time before checking if interrupt handler has been executed.

mprse committed:
lp_us_tickers test - fix count_ticks() function to give the same results on all compilers

count_ticks() function counts ticker ticks elapsed during execution of N cycles of empty while loop.
In current version N value (cycles) is given as volatile paramater in order to disable possible compiler optimalisation.
There was a problem with measured time on different compilers and additionally results on ARM compiler were unexpected (the difference beetween measured elapsed ticks for the same number of cycles was to large). This might be caused by the memory access in order to store updated variable in memory. To fix this issue given numer of cycles has been stored into register and register is decremented (no memory access).

With this fix count_ticks(NUM_OF_CYCLES, 1) call returns 2500 +/-1 for us ticker ticks using each compiler (K64F).

mprse committed:
Enable RTC support for FAMILY_STM32.

mprse committed:
HAL rtc test: check only if rtc_isenabled() returns 1 in case when init is done.

Since some platforms use RTC for low power timer RTC may be already enabled.
Because of that we will only verify if rtc_isenabled() returns 1 in case when init is done and RTC time is set.

mprse committed:
HAL rtc test: increase test timeout.

Verified that on some platforms 30s is not enough.

mprse committed:
Skip higher level ticker tests for targets with stale ticker target specific drivers.

Since target specific ticker drivers are not ready also features which uses ticker in upper layers may not work correctly and tests for these features.
We need to disable also failing higher level ticker related tests (by adding check if DEVICE_USTICKER symbol is available and raise #error [NOT_SUPPORTED] if not).

Note:
tests-mbed_drivers-rtc is new and uses us ticker to perform a delay.

mprse committed:
lp_ticker test - provide lp ticker glitch test case

Test that lp ticker does not glitch backwards due to an incorrectly implemented ripple counter driver.

mprse committed:
lp_us_tickers test: fix `increment test` for slow boards

Increment test proves that ticker counter is incremented by one.
This is done indirectly for high frequency counters where it is impossible to read 'value' and 'value + 1' in two successive ticker reads.
This check is done indirectly by counting ticker ticks elapsed during execution of N cycles of empty while loop. Unfortunately on slow boards with fast tickers like NRF51_DK(16 MHz CPU/1MHz hf ticker) it is possible that for the same N cycles measured number of elapsed ticks in two successive calls is greater than 1. This patch provides fix for such case - measure operation is repeated with the same number of cycles.

mprse committed:
lp_us_tickers test: fix overflow protection

On slow targets with fast high frequency tickers like NRF51_DK (16 Mhz CPU/1MHz ticker) time window for test case execution without overflow needs to be increased. Add parameter to `overflow_protect` function to be able to set different time window without overflow for us ticker and lp ticker.

mprse committed:
Enable RTC support for K64F board.

mprse committed:
K64F: adapt RTC drivers to the new standards (free function)

rtc_free() does not stop the RTC from counting. Only disables clock gate since processor no longer needs to read RTC registers.

mprse committed:
K64F: adapt RTC drivers to the new standards (is_enabled function)

Make rtc_isenabled() to return 1 if the RTC is initialized and the time has been set; 0 otherwise.
Disable clock gate on exit from this function if RTC was initialized.

mprse committed:
tests-mbed_drivers-timerevent: perform test only if USTICKER support is enabled.

mprse committed:
tests-mbed_hal-lp_ticker:  run sleep test only if SLEEP support is enabled.

mprse committed:
tests-mbed_hal-rtc_reset: Add ack from the device after each command is executed

RTC reset test was failing when board has been just powered and RTC reset test was executed for the first time (issue has been detected on CI). In such case RTC initialization takes more time than in futher rtc_init calls. This has impact on green-tea communication. Commands send by host immediately after init command (write, read) were not handled on the device side and no response to the host was provided.
To fix this problem test communication flow has been modified. Device sends ack to the host after RTC init is done. Host sends further RTC commands when ack from the device has been received.

Edit:
There are still some communication issues on the CI. Add ack from the device after each executed command. Increase test timeout and wait after reset.

mprse committed:
Disable RTC support for UBLOX_EVK_ODIN_W2.

This board inherits from STM_FAMILY, but drivers for this board are not consistent with the new standards.
This board needs to be disabled for now.

mprse committed:
lp_us_tickers test: provide minor fixes after review

- count_ticks: fix counter overflow handling,
- count_ticks: use reg_cycles variable in while loop,
- increment test: reduce number of cycles for slow cores if measure process needs to be repeated (difference is greater than 1).

mprse committed:
tests-mbed_drivers-lp_ticker: Adapt tolerance for new Ticker driver

Since this change concerns new ticker driver for NRF51_DK it goas on ticker feature branch.

On master NRF51_DK Ticker driver uses 32kHz RTC for lp and us ticker. Test uses us ticker to measure lp ticker interrupt scheduling (LowPowerTicker.attach), but since the same hardware is used for both tickers the measurement error is constant. On this branch us ticker uses 1 MHz higher precision clock and the results are different - measurement error grows linearly and shows inaccuracy of the RTC.
Test implements constant delta for measured time equal to 2000 us.

Change delta so it depends on lp ticker tested timeout value:
500us for measurement inaccuracy + 5% tolerance for LowPowerTicker

mprse committed:
tests-mbed_drivers-timer: Adapt tolerance for new Ticker driver

Since this change concerns new ticker driver for NRF51_DK it goas on ticker feature branch.

On this branch us ticker is based on 1 MHz higher precision counter (on master is based on 32kHz RTC). As a result of using higher precision counter measurement error increased:
:485::FAIL: Expected 1.060000 Was 1.070478

For delay 1060 ms we measured ~1070 ms so we have about ~1% measurement error which is not bed for this board.

To make it work with 1MHz us ticker increase test tolerance. Define tolerance as follows:
tolerance  = 500us + 0.02 * delay.

bulislaw committed:
Rename DEVICE_LOWPOWERTIMER to DEVICE_LPTICKER

That's to match DEVICE_USTICKER.

bulislaw committed:
Disable Maxim boards as LP_TICKER is used in RTC and they don't respect device_has

bulislaw committed:
Make sure LPTICKER symbols are not used for builds without it for Nordic

bulislaw committed:
Don't use tickless if LPTICKER is not present

The current implementation of tickless requires LPTICKER to be present.

bulislaw committed:
Disable platfroms that fail without LPTICKER

bulislaw committed:
Remove nrf51_dk from supported platform

bulislaw committed:
Rename lp_us tests to common and hf to microsecond

LMESTM committed:
Update STM32 16 bits us_tickers in line with new mbed HAL

The new HAL allows to share the timer bit width and frequency,
the actual handling of mapping 16 bits counter up to 32 bits or
64 bits is now managed by mbed common layer.

This makes this ticker layer very similar to 32bits one and much
easier than before.

mmahadevan108 committed:
MCUXpresso: Enable RTC support

Same MCUXpresso SDK driver used for Kinetis devices

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

LMESTM committed:
Make us_ticker common between 16 and 32 bits counters

mmahadevan108 committed:
MCUXpresso: Enable RTC on LPC54114 and LPC546XX

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

stevew817 committed:
Re-implement RTC for Silicon Labs targets

mbed RTC specifications now dictate that the RTC needs to retain and keep on counting through reset. On Silicon Labs parts, this means the RTC API can not be backed by the Silicon Labs RTC peripheral, since that doesn't provide retention functionality.
Therefore:
* On EFM32GG, EFM32WG, EFM32LG: mbed RTC API is now backed by BURTC.
* On EFM32PG, EFR32MG, EFM32PG12, EFR32MG12: mbed RTC API is now backed by RTCC.
* On EFM32ZG, EFM32HG: mbed RTC API is sadly no longer supported, since these chips don't have retained memory.

stevew817 committed:
Re-implement RTC for Silicon Labs targets

mbed RTC specifications now dictate that the RTC needs to retain and keep on counting through reset. On Silicon Labs parts, this means the RTC API can not be backed by the Silicon Labs RTC peripheral, since that doesn't provide retention functionality.
Therefore:
* On EFM32GG, EFM32WG, EFM32LG: mbed RTC API is now backed by BURTC.
* On EFM32PG, EFR32MG, EFM32PG12, EFR32MG12: mbed RTC API is now backed by RTCC.
* On EFM32ZG, EFM32HG: mbed RTC API is sadly no longer supported, since these chips don't have retained memory.

# Conflicts:
#	targets/TARGET_Silicon_Labs/TARGET_EFM32/lp_ticker.c
#	targets/TARGET_Silicon_Labs/TARGET_EFM32/rtc_api.c
#	targets/targets.json

mprse committed:
lp_us_tickers test: call overflow_protect function in test functions instead test setup handler.

This is done to bypass green-tea setup handler which takes additional time.

bulislaw committed:
Fix rebase error

stevew817 committed:
Re-implement us_ticker and lp_ticker for Silicon Labs targets

Re-implemented both us_ticker and lp_ticker to match the new API and specifications.
Details:
* On EFM32GG, EFM32WG, EFM32LG, EFM32HG, EFM32ZG: Use the RTC peripheral to back lp_ticker, and a TIMER to back us_ticker.
* On EFM32PG, EFR32MG, EFM32PG12, EFR32MG12: Use the RTCC peripheral to back lp_ticker (dual-purpose, also used to back RTC), and a TIMER to back us_ticker.

jeromecoutant committed:
STM32 LPTICKER update for targets supporting LPTIMER

jeromecoutant committed:
LPTICKER targets json update for STM32 supporting LPTIMER

STM32L0, STM32L4, STM32F7 and few STM32F4

mmahadevan108 committed:
MCUXpresso: Update Low Power Ticker

1. Enable LPTICKER for K22, K24, K64, K66, K82, KL82F, KW24D
2. Change the implementation to only use the LPTMR which reduces
   the amount of interrupts generated which is required for tickless
   operation

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

stevew817 committed:
Code style fixes requested by @0xc0170

Tests still passing

mprse committed:
tests-mbed_hal-sleep: add lp/us ticker overflow handling.

mprse committed:
tests-mbed_hal-sleep: use lp ticker data while testing deepsleep.

mprse committed:
tests-mbed_hal-sleep: remove unused variable.

mprse committed:
tests-mbed_hal-sleep: fix formatting.

mprse committed:
tests-mbed_hal-sleep: decrease delay when testing if us ticker is disabled in deep-sleep mode.

This is done because of possible limitations of lowpower ticker freq/width.

mprse committed:
Rename DEVICE_LOWPOWERTIMER to DEVICE_LPTICKER.

mprse committed:
ARM_CM3DS_MPS2: add protection to lp_ticker.c.

Compile this file only when DEVICE_LOWPOWERTIMER is defined.

jeromecoutant committed:
STM32 LPTICKER : enable all STM32 targets supporting the feature

(All STM32 except F1 and F2 families)

jeromecoutant committed:
STM32 LPTICKER : clean include file

TomoYamanaka committed:
Implement of RTC feature for Renesas mbed boards

I implemented the RTC feature.
The mainly changing is here.
- rtc_init()
  Previously, I have initialized the time information register in the function, so the time count was cleaned by every calling rtc_init().
  Currently, rtc_init() doesn't stop RTC from counting, and rtc_init() is safe to call repeatedly.
  Therefore in order to satisfy specifications,I removed the initialization process of the time information register in the function.

- rtc_free()
  Previously, I have initialized the RTC related register same as rtc_init(), so the time count was cleaned by calling rtc_free().
  Currently, rtc_free() doesn't stop RTC from counting.
  Therefore in order to satisfy specifications,I removed the process and decided not to do anything in the function.
  If powerdown the RTC, Supply of the clock to the RTC is stopped, cannot keeping the count.

TomoYamanaka committed:
Implementation of SLEEP feature for Renesas mbed boards

I implemented the SLEEP feature for Rnesas mbed boards.
The mainly changing is here.
- hal_sleep()
  To satisfy the mbed specificationm, I implemented this function newly by using "sleep" that is one of low power mode that is incorporated in our hardware(RZ_A1).
  In the "sleep", peripheral and memory state are maintained, and the peripherals continue to work and can generate interrupts.

- hal_deepsleep()
  To satisfy the mbed specificationm, I implemented this function newly by combined using "sleep" and "module standby" that is the low power mode that is incorporated in our hardware(RZ_A1).
  The "module standby" is peripheral module's powerdown.
  Also in case of our "module standby", it need to read register as dummy when access to each register.

TomoYamanaka committed:
Implementation of USTICKER feature for Renesas mbed boards

I implemented USTICKER feature.

The mainly changing is here.
- I added a macro to mbed_drv_cfg.h for commonalizing code for GR-PEACH and GR-LYCHEE with different clock frequencies, and referenced it's macro at us_ticker.c.
- ticker_init()
  Currently, ticker_init() keep counting, disables the ticker interrupt, and is safe to call repeatedly.
  Therefore, in order to satisfy specifications, I removed GIC_EnableIRQ at end of function and added GIC_DisableIRQ at begin of function.
  When an interrupt is required, it will be set with ticker_set_interrupt().
  If executing the following, the counter has been initialized. So it will not call after executing the first time.
    OSTM1TT   = 0x01;    /* Stop the counter and clears the OSTM1TE bit.     */
    OSTM1TS   = 0x1;    /* Start the counter and sets the OSTM0TE bit.     */

- ticker_free()
  this function stops the counting and powerdown the us_ticker.
  To satisfy the mbed specificationm, I implemented free() function.

- ticker_read()
  Currently, Mbed spec's frequeny is between 250KHz and 8MHz, but the frequency that is used at my ticker is 33MHz.
  Therefore, in order to satisfy specifications, I changed the process to return the timer counter value divided by 32(33MHz / 32).
  Since the calcurate function by using 64 bit is no longer necessay, I removed it.

- ticker_set_interrupt()
  Same as the above read(),
  In order to satisfy specifications, I changed the process to set the value multiplied by 32.

- ticker_fire_interrupt()
  In order to satisfy specifications, I implemented fire_interrupt() function.
  Also I added GIC_EnableIRQ for allowing the interrupt at end of function.

- ticker_get_info()
  To satisfy the mbed specificationm, I implemented ticker_get_info() function. The value of freq includes rounding off.

mprse committed:
Disable SLEEP support for MAX32630FTHR and KW41Z platforms.

These platforms are not consistent with the new sleep standards.

mprse committed:
tests-mbedmicro-rtos-mbed-systimer: execute sleep/deepsleep test cases only when SLEEP support is enabled.

mprse committed:
Disable `tests-mbed_drivers-timerevent` test if USTICKER support is disabled.

mprse committed:
Disable `tests-mbed_drivers-timeout` test if USTICKER support is disabled (NRF52_DK).

c1728p9 committed:
Restore ticker state after each test

Update the ticker common test to clean up after each case by restoring
the ticker IRQ handler. This allows tickers to function normally after
each case has completed.

jeromecoutant committed:
tests-mbed_hal-common_tickers_freq : correct overflowCounter value

jeromecoutant committed:
tests-mbed_hal-common_tickers_freq : use unsigned long

jeromecoutant committed:
tests-mbed_hal-lp_ticker : increase time before deep sleep

mprse committed:
tests-mbed_hal-lp_ticker: Add overflow protection.

jeromecoutant committed:
STM32 LPTICKER : read counter

bulislaw committed:
Disable platforms not supporting new RTC

bulislaw committed:
Rename LOWPOWERTIMER to LPTICKER after merge

mprse committed:
tests-mbed_hal-common_tickers: increase overflow protection limit

For NR51_DK US_TICKER_OV_LIMIT needs to be increased since if test is run few times in row sometimes fails. This is because NR51_DK is a slow board (16 MHz) with fast and short us ticker counter 1 MHz/16 bits.

jeromecoutant committed:
#6536 rebase correction

jeromecoutant committed:
STM32 LPTICKER update for targets supporting RTC

jeromecoutant committed:
STM32 RTC Init minor update

Commit:
mbed_ticker_api remove specific case for 32768 in favor of all frequencies divisible by 2

Commit:
Changed struct position to try to maintain binary compatibility
Changed comment because a lot of numbers are divisible by 2

Commit:
Updated types according to header changes

bulislaw committed:
Disable failig Nordic targets

mmahadevan108 committed:
MCUXpresso: Update the usticker implementation

Enabled usticker for K22F, K24F, K66F, K82F, KL82Z, KW24D

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
MCUXpresso: Enable usticker for LPC546XX and LPC54114

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
MCUXpresso: Enable usticker on MIMXRT1050_EVK

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

bulislaw committed:
Remove sleep from rf52 due to failig tests

bulislaw committed:
Disable MAX targets failig to build

mprse committed:
Move common_rtc.c and us_ticker.h files to the valid directory.

I decided to move these files to the targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF51 since us_ticker.h is for sure specific for NRF51_DK and common_rtc.c might be valid also for NRF52, but this needs to be checked while porting NRF52_DK board.

mprse committed:
Enable support for NRF51 boards.

mprse committed:
tests-events-queue: modify tolerance to percentage form

On some boards without SYSTICK support lp ticker is used to trace system ticks.
Event queue uses system ticker to count delay when callback is to be fired.
New ticker implementation for NRF51_DK provides fast and accurate counter for us ticker which shows inaccuracy of the system ticker(which is based on slow and inaccurate lp ticker). This is why measured error grows linearly and percentage form needs to be used.

Currently we use constant tolerance value equal to: 5 [ms].
This patch modifies tolerance to the following form: 5% of measured time [ms].

In this test the lowest delay is equal to 100 [ms], so 5% corresponds to 5 [ms] tolerance - original value. This means that min tolerance is 5 [ms] so this change should not have negative impact on other targets.

mmahadevan108 committed:
KW41Z: Update usticker to use TPM instead of LPTMR for ticker

This will free up the LPTMR to be used as lpticker

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
MIMXRT1050_EVK: Update lpticker implementation

Use only the GPT module and avoid using RTC.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mprse committed:
tests-mbed_drivers-timerevent: Disable deep-sleep mode in TestTimerEvent class.

Since now us ticker is disabled in deep-sleep mode and tests-mbed_drivers-timerevent verifies TimerEvent class which uses us ticker then deep-sleep mode needs to be disabled during the test execution.

mprse committed:
tests-mbed_drivers-lp_ticker: increase tolerance for NRF51_DK

This test fails sometimes while testing 1 ms timeout on NRF51_DK board.
This board is very slow and lp ticker is inaccurate.
I noticed that measured results are very close to the tolerance value. For 1ms delay we measure in failure case 1560 us and tolerance is set to 550.
Since measurement method is also significantly flawed (extra time for setting up ticker and execution of a ticker irq handler) I recommend to increase tolerance by least +100 us for now.

mprse committed:
Disable us ticker in deep-sleep mode.

mmahadevan108 committed:
KL05, KL25, KL26, KL46: Enable usticker

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
LPC1768: Enable usticker

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

jessexm committed:
mbed-os v5.9 RTC implementation

jessexm committed:
mbed-os v5.9 USTICKER implementation

jessexm committed:
mbed-os v5.9 LPTICKER implementation

Commit:
ADI: Re-enable DEVICE_RTC for EV_COG_AD3029LZ and EV_COG_AD4050LZ

mprse committed:
Enable support for NRF51 boards.

Make it consistent with master.

mprse committed:
tests-mbed_hal-common_tickers: disable os tick while running test cases

It has been noticed on NRF51_DK board that occurrence of system tick have impact on test execution.
NRF51_DK is a slow board with fast us ticker and handling of system tick interrupt takes about 250 us ticker ticks which have huge influence on interrupt tests where interrupt is set to <current count> + 100, 200, 500 ticks.

Commit:
Remove trailing spaces of adi_rtc.c

Commit:
Preserve RTC counter value while re-initialization

mmahadevan108 committed:
KL43Z: Update usticker to use TPM instead of LPTMR for ticker

This will free up the LPTMR to be used as lpticker

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
KL27Z: Update usticker to use TPM instead of LPTMR for ticker

This will free up the LPTMR to be used as lpticker

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
KW41Z, KL43Z, KL27Z: Enable LPTICKER

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

Commit:
Fix us_ticker.c to support "USTICKER"

- Disable microsecond ticker interrupt on reinitialization
- Skip us_ticker_set_interrupt() if timestamp is already past
- Eliminate tmr2Config since tmrConfig is adequate for all timer config

mprse committed:
NRF52_DK: use new ticker driver designed for NRF51_DK

mprse committed:
NRF52_DK: Enable lp ticker and us ticker support

mprse committed:
tests-mbed_hal-common_tickers: fix bug in ticker increment test case

It is possible that the difference between base and next tick count on some platforms is greater than 1, in this case we need to repeat counting with the reduced number of cycles (for slower boards).
If number of cycles has been reduced than base tick count needs to be redefined. This operation is missing and is added by this patch.

mprse committed:
tests-mbed_drivers-timerevent: increase tolerance for NRF52_DK

Test inserts event into the TimerEvent object which is scheduled 50 ms in the future. Then test thread hangs on the semaphore for 51 ms and after that time we expect that event handler has been executed. This test fails sometimes on NRF51_DK, NRF52_DK since different clocks are used for event handling and delay. TimerEvent class uses fast us ticker and semaphores are based on lp ticker (NRF51_DK) or System Tick Timer (NRF52_DK).
We assume that ticker measurement error is 10% and our 1 [ms] extra corresponds to 2% error. I suggest to increase delay to 2 [ms] (4%). This should be enough for all boards at this moment.

mprse committed:
tests-mbed_drivers-lp_timer: modify tolerance to percentage form

Represent tolerance value as it is done in `tests-mbed_drivers_timer`.
Use 5% tolerance for lp ticker.

mprse committed:
tests-mbed_hal-flash: Remove 5% extra tolerance for NRF52_DK

NRF52_DK is now based on fast and accurate 1MHz counter.

mprse committed:
tests-mbed_hal-flash: optimieze time_cpu_cycles() function

This function should perform instruction cycles count in critical section.
Additionally remove redundant code.

mprse committed:
NRF52_DK: disable us ticker in deep-sleep mode

This functionality is required by new sleep standards.

mprse committed:
Make common_rtc.c compatible with SDK 14.2 and move it common dir

mprse committed:
Enable us lp ticker support for entire MCU_NRF52832 family

bulislaw committed:
Change LPTICKER spec to allow clock freq starting at 4kHz

Commit:
Adding usticker to MTB_ODIN

mprse committed:
NRF52840_DK: Enable lp us ticker support

bulislaw committed:
Rename new occurrences of LOWPOWERTIMER to LPTICKER

bulislaw committed:
Fix missmatch between feature branches and new CMSIS

bulislaw committed:
Fix target definition for Odin boards

bulislaw commented:
@ARMmbed/mbed-os-maintainers please run ci.
@ithinuel @jamesbeyond @0xc0170 could you keep an eye on that and try to fix it if somethings goes wrong.

0xc0170 commented:
/morph build

0xc0170 reviewed (dismissed):
What are the coverage numbers for these features (how much target coverage percentage ?) ?

0xc0170 commented:
Reviewing build failures now

0xc0170 committed:
maxim: fix rtc if lp ticker not defined

lp ticker handler defined only if lp ticker is defined. In case not,
use only overflow, rtc0 not used.

0xc0170 committed:
efm32: remove unused undefined function from rtc

0xc0170 commented:
Resolved both errors, tested locally , should build now and all be good 🤞 

/morph build

0xc0170 committed:
max32600: fix lp ticker functions - do not inline

Wrong linkage specified, causing build to fail.

0xc0170 commented:
One error missed (gcc somehow passed 😕 ), restarting

/morph build

ithinuel commented on TESTS/mbed_drivers/lp_timer/main.cpp:
```diff
@@ -35,26 +35,21 @@ extern uint32_t SystemCoreClock;
+ *  */
+#define DELTA_US(delay_ms) (500 + delay_ms * US_PER_MSEC / 20)
+#define DELTA_MS(delay_ms) (1 + (delay_ms * US_PER_MSEC / 20 / US_PER_MSEC))
+#define DELTA_S(delay_ms) (0.000500f + (((float)delay_ms) / MSEC_PER_SEC / 20))
```
As macro in C (and c++) are basic text subtitution, it is usually good practice to wrap the macro's argument between brackets.
If you pass to `DELTA_US` an expression such as `a_var + something_else` you may not obtain what you expect.
```
#define DELTA_US(delay_ms) (500 + (delay_ms) * US_PER_MSEC / 20)
```

Also, as this is in test it does not really impact the final code but in debug mode, double division may not get optimised away and  `(1 + (delay_ms * US_PER_MSEC / 20 / US_PER_MSEC))` will try to run 2 divisions.

ithinuel replied:
From what I see on GH these three macros are defined twice (the other definitions are line 63).

adbridge replied:
And worse than that they are actually defined differently!!

mprse replied:
Nice catch. I agree with the first part, but since arithmetic operations are not passed to the macro its not a critical error.
Unfortunately the second bug is unsafe. It looks like I forgot to remove old definitions 😢 .

0xc0170 replied:
Can you add a commit fixing this?

mprse replied:
I don't know how I can add commit to `bulislaw:merge_feature_branches_for_5.9` branch.
Lines 49-51 should be inserted instead of code in lines to 59-65.

But the old version is not bed (and tested). It has been changed for consistency with other ticker tests.

ithinuel replied:
@bulislaw needs to add you as collaborator to his repo.

0xc0170 replied:
I can help, got that branch locally, point me to a commit and I'll cherry pick and push

mprse replied:
Fix can be found here:
https://github.com/mprse/mbed-os/commit/83a0146b88462b8fdad2ef06964fdf60eaf9d899

Tested on `NRF51_DK`, `NRF52_DK` boards and `GCC_ARM`.

ithinuel replied:
@mprse actually there are arithmetic in [mbed_drivers/timer/main.cpp#L724](https://github.com/bulislaw/mbed-os/blob/085a3a08a2793b2302ac77292e63734c8aedbe2d/TESTS/mbed_drivers/timer/main.cpp#L724)
It is ok as it is just division but for maintainability it might create non-obvious bugs.

mprse replied:
Provided fix adds brackets.

ithinuel reviewed (changes requested):
This PR is pretty big so I'll slip my reviews so I don't loose pending comments.

mprse committed:
Fix tolerance definition in ticker tests.

ithinuel commented on TESTS/mbed_hal/common_tickers/ticker_api_tests.h:
```diff
@@ -0,0 +1,129 @@
+ *
+ * We have the following assumption for the timer clock frequency:
+ *
+ * NOTE:
```
To be generated as an actual note block this needs to be written `@note` or `\note`

ithinuel commented on hal/lp_ticker_api.h:
```diff
@@ -30,10 +30,37 @@ extern "C" {
  */
 
+/**
+ * \defgroup hal_lp_ticker_tests Low Power Ticker tests
```
@SenRamakri recently pointed that this should probably not be captured in the doc.
Is there any update ?

c1728p9 replied:
There needs to be details on how to run hal testing. Is @SenRamakri proposing this gets moved or removed?

mprse commented:
I have noticed that there is some inconsistency on this branch `DEVICE_LOWPOWERTIMER` should be replaced with `DEVICE_LPTICKER` in all source files and tests.

mprse commented:
`tests-mbed_hal-rtc_reset` has failed on NRF52_DK board (all compilers), but as far as I remember RTC support should be disabled for this board (since it can not persist time over the reset).

mprse commented:
`tests-mbed_hal-sleep` has failed on NRF51_DK board (all compilers). Only one test case from three has been executed (because of `DEVICE_LOWPOWERTIMER` issue). The test which checks if us ticker can wake up board from sleep has failed. Don't know why - can't reproduce locally (test passes - also with additional CI flags). Need to investigate - maybe it is caused by `DEVICE_LOWPOWERTIMER` issue.

fkjagodzinski commented on TESTS/mbed_drivers/timerevent/main.cpp:
```diff
@@ -23,7 +27,12 @@
 
+#if !DEVICE_USTICKER
+  #error [NOT_SUPPORTED] test not supported
+#endif
```
Lines 30-32 duplicate lines 16-18 of this file.

fkjagodzinski reviewed (commented):
Just one minor detail.

bulislaw committed:
Rename LOWPOWERTIMER to LPTICKER

bulislaw committed:
NRF52 is not compatible with new RTC HAL spec

ithinuel reviewed (commented):
In addition to the comments I noticed that Copyright headers' date is not written/maintained consistently.
shouldn't it be kept as `Copyright (c) file_creation_year-current_year ARM Limited` ?

bulislaw committed:
Sleep was not ported to MAX32630FTHR

bulislaw committed:
Fix missmatch between targets.son

cmonr commented:
Afaik, all comments have been addressed with the rebase + updates. 

@ithinuel I've forwarded the question to @ChiefBureaucraticOfficer who has forwarded it to legal. Something like that can be fixed for RC2.

Gonna start CI!
/morph build

cmonr commented:
Quick summary:
NUCLEO_F746ZG
 - tests-mbed_hal-sleep 
 - All compilers

KW41Z
- tests-mbed_hal-sleep 
- GCC & IAR

MAX32630FTHR 
- tests-mbed_hal-rtc
 - All compilers

EV_COG_AD3029LZ
- tests-mbed_hal-flash
- IAR and ARM

Additionally, NRF51 timed out, but was also failing tests. Because build was auto-aborted due to timeout, logs were not pushed, but are still available in CI.
NRF51_DK
- tests-mbed_hal-sleep
- All compilers.

cmonr commented:
Rerunning test build since all is quiet...
Not expecting different results.

/morph test

bulislaw commented:
Weirdly the results are different...

bulislaw committed:
Disable RTC on MAX32630FTHR due to failing tests

bulislaw committed:
Disable intermittently failing sleep test

bulislaw committed:
Disable FLASH for AD boards as they fail the tests intermittently

bulislaw commented:
I've disabled RTC on MAX32630FTHR and disabled the sleep wake up test, as it's failing intermittently on different boards. I don't know what's wrong with flash tests on AD boards, but it seems to be failing intermittently, therefore removing FLASH capabilities from these boards for now.

cmonr commented:
Iirc, the Flash Cache test had some sort of fix applied in master and the problem went away.

Something to look into next week.

/morph build

bulislaw commented:
I've rebased it on top of master yesterday so it should pick up all the fixes, maybe it's unstable. I'll put an email together with all the things to check/address during OOB.

cmonr commented:
Should, yes. But if the API/mechanism of the failure changed, it's possible that it was/could have been reintroduced.

bulislaw committed:
Disable FLASH algo for AD boards

bulislaw commented:
Fiddling with the flags for AD...

/morph build

mprse commented:
I see that we have general problems with sleep tests.  I think that it is possible that unstability of this test is cased by system tick. I saw that on ticker tests. We can try to disable systick during the sleep tests.

@bulislaw it looks like you have disabled wrong test case in the sleep test.

bulislaw commented:
/morph build

bulislaw commented:
/morph build

bulislaw commented:
@cmonr Could you merge 2 PR for the exampels and start tests here. BTW there's a bug in the CI, exporters do not respect `[NOT SUPPORTED]` error and fail the build rather than skip. @studavekar

cmonr commented:
Interesting.

Unfortunately, since @deepikabhavnani is the owner of those two example repos, so we might be done for the weekend in trying to get this in. I've sent her a PM for now, but since it's technically a holiday weekend, we might be stuck.

bulislaw commented:
/morph build

0xc0170 commented:
> Unfortunately, since @deepikabhavnani is the owner of those two example repos, so we might be done for the weekend in trying to get this in. I've sent her a PM for now, but since it's technically a holiday weekend, we might be stuck.

As I can see those 2 are merged. Outstanding here is to make tests green (exporters green). I got some of those devices, will check tomorrow morning the failures

bulislaw committed:
Disable sleep tests as they are unstable

bulislaw commented:
/morph build

@0xc0170 @mprse I've disabled the sleep tests as both lp and us variant are failing for multiple boards, could you investigate on Monday? In the meantime if the testing passes we can merge it and fix these and more issues during RC, I'll send separate email detailing things to look at.

cmonr commented:
@0xc0170 Sorry, but to mention that Deepika came online and merged those two PRs, unblocking this one.

bulislaw commented:
/morph build

mprse commented:
Looking into the sleep issue.

0xc0170 commented:
@ithinuel Were your comments addressed ? If not, are they blocking ?

ithinuel commented:
@0xc0170 One is awaiting an answer from @SenRamakri but I don't think they are blocking
Do you need me to change the review's status ?

ithinuel reviewed (approved):
The documentation related comments can be addressed later.

mprse commented:
Fix for http://mbed-os-logs.s3-website-us-west-1.amazonaws.com/?prefix=logs/7009/1959/FAIL/NRF51_DK/ can be found here:

https://github.com/mprse/mbed-os/commit/eeab68d44cdd6e3549e97b5bed9be0a19232cd3c
